docs/prepare-db-upgrade.md

```diff
@@ -0,0 +1,135 @@
+
+## Details
+
+An `upgrade.properties` file will look something like:
+
+```
+description: what it does
+compatibility: partial
+some_relation.rel: run some_relation.qlo
+```
+
+The `description` field is a textual description of the aim of the upgrade.
+
+The `compatibility` field takes one of four values:
+
+ * **full**: results from the upgraded snapshot will be identical to results from a snapshot built with the new version of the toolchain.
+
+ * **backwards**: the step is safe and preserves the meaning of the old database, but new features may not work correctly on the upgraded snapshot.
+
+ * **partial**: the step is safe and preserves the meaning of the old database, but you would get better results if you rebuilt the snapshot with the new version of the toolchain.
+
+ * **breaking**: the step is unsafe and will prevent certain queries from working.
+
+The `some_relation.rel` line(s) are the actions required to perform the database upgrade. Do a diff on the the new vs old `.dbscheme` file to get an idea of what they have to achieve. Sometimes you won't need any upgrade commands – this happens when the dbscheme has changed in "cosmetic" ways, for example by adding/removing comments or changing union type relationships, but still retains the same on-disk format for all tables; the purpose of the upgrade script is then to document the fact that it's safe to replace the old dbscheme with the new one.
+
+Ideally, your downgrade script will perfectly revert the changes applied by the upgrade script, such that applying the upgrade and then the downgrade will result in the same database you started with.
+
+Some typical upgrade commands look like this:
+
+```
+// Delete a relation that has been replaced in the new scheme
+obsolete.rel: delete
+
+// Create a new version of a table by applying an expression (using a simple
+// synthetic language) to an existing table. The example duplicates the 'id'
+// column of input.rel as the last column of extended.rel, perhaps to record our
+// best guess at newly-populated "source declaration" information.
+extended.rel: reorder input.rel (int id, string name, int parent) id name parent id
+
+// Create relationname.rel by running relationname.qlo and writing the query
+// results as a .rel file. The query file should be named relationname.ql and
+// should be placed in the upgrade directory. It should avoid using the default
+// QLL library, and will run in the context of the *old* dbscheme.
+relationname.rel: run relationname.qlo
+```
+
+### Testing your scripts
+
+Although we have some automated testing of the scripts (e.g. to test that you can upgrade databases all the way from the an initial dbscheme to the newest, and back), it's essential that you apply some more rigorous testing any non-trivial upgrade or downgrade. You might do so as follows:
+
+#### Running qltests
+
+To test the upgrade script, run:
+
+```
+codeql test run --search-path=<old-extractor-pack>;ql <test-dir>
```

Comment:
`;` is the windows path separator. In addition, I think you no longer need to specify `ql` in the search-path
```suggestion
codeql test run --search-path=<old-extractor-pack> <test-dir>
```

Reply:
When I try without `:ql`, I get

```
Could not evaluate queries in .../ql/test/library-tests/ast/control: The CodeQL database is not compatible with a QL library  referenced by the query you are trying to run.
	 CodeQL database location: ql/test/library-tests/ast/control/control.testproj
	Query location: ql/test/library-tests/ast/control/CaseExpr.ql
The database may be too new for the QL libraries the query is using; try upgrading them.
```

Reply:
Ah ok, that's strange. `codeql` is able to compile the test query which (transitively) depends on `codeql/ruby-all`, so I wonder why it cannot find the upgrade scripts that are inside that pack without adding a search path. @aeisenberg do you understand this?

Instead of using `;` I think you can also repeat the `--search-path` flag .